libnfs4acl/nfs4_ace_to_json.c

```diff
@@ -0,0 +1,357 @@
+ *
+ * Copyright (c) 2021 iXsystems, Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <stdio.h>
+#include <ctype.h>
+#include <assert.h>
+#include <jansson.h>
+#include "nfs4_json.h"
+#include "libacl_nfs4.h"
+
+static int
+who_to_json(json_t *_parent, struct nfs4_ace *entry, bool numeric)
+{
+	int error;
+	char *tag = NULL;
+	char who_str[NFS4_MAX_PRINCIPALSIZE] = {0};
+	int who_id = -1;
+
+	switch (entry->whotype) {
+	case NFS4_ACL_WHO_NAMED:
+		tag = NFS4_IS_GROUP(entry->flag) ? "group" : "user";
+		if (numeric) {
+			error = acl_nfs4_get_who(entry, &who_id, NULL, 0);
+			if (error) {
+				return (error);
+			}
+		}
+		else {
+			error = acl_nfs4_get_who(entry, &who_id, &who_str,
+						 NFS4_MAX_PRINCIPALSIZE);
+			if (error) {
+				return (error);
+			}
+		}
+		break;
+	case NFS4_ACL_WHO_OWNER:
+		tag = "owner@";
+		break;
+	case NFS4_ACL_WHO_GROUP:
+		tag = "group@";
+		break;
+	case NFS4_ACL_WHO_EVERYONE:
+		tag = "everyone@";
+		break;
+	default:
+		return (-1);
+
+	}
+	error = json_object_set_new(_parent, "tag", json_string(tag));
+	if (error) {
+		return (error);
+	}
+	if (!numeric && entry->whotype == NFS4_ACL_WHO_NAMED) {
+		error = json_object_set_new(_parent, "who", json_string(who_str));
+		if (error) {
+			return (error);
+		}
+	}
+	error = json_object_set_new(_parent, "id", json_integer(who_id));
+	if (error) {
+		return (error);
+	}
+
+	return (0);
+}
+
+static int
+perms_to_json(json_t *_parent, nfs4_acl_perm_t access_mask, bool verbose)
+{
+	int error, i;
+	json_t *basic = NULL, *perms = NULL;
+	perms = json_object();
+	if (perms == NULL) {
+		return (-1);
+	}
+	if (!verbose) {
+		if (access_mask == (NFS4_ACE_FULL_SET)) {
```

Comment:
Could be a switch instead. It also seems like these sets could be useful to define in a header somewhere, if they're interesting enough to be exposed to users this way.